ivc/src/ivc/columns.rs

```diff
@@ -0,0 +1,141 @@
+///      |  ...??? empty space?
+///      |     reuse w/ selectors?
+/// 2^15 |---------------
+///```
+///
+/// Counting cells:
+/// - Inputs:              2 * 17 * 3N = 102N
+/// - Inputs repacked 75:  2 * 4 * 3N = 24N
+/// - Inputs repacked 150: 2 * 2 * 3N = 12N
+/// - Hashes:              2 * 165 * 3N = 990N (max 4 * 165 * 3N if we add 165 constants to every call)
+/// - scalars:             4 N + 17 * 3 * N = 55 N
+/// - ECADDs:              230 * 35 * 3N = 24150N
+/// Total:                 25333*N
+///
+///     ...which is less than 32k*N
+///
+/// We can calculate N_IVC as dependency of N_APP in this way:
+///    N = N_APP + (CELL/2^15)*N
+///    (1 - CELL/2^15)*N = N_APP
+///    N = (1/(1 - CELL/2^15)) * N_APP = (2^15 / (2^15 - CELL)) * N_APP
+///    N_IVC = (1/(1 - CELL/2^15) - 1) * N_APP = (2^15 / (2^15 - CELL) - 1) * N_APP
+///
+/// In our particular case, CELL = 25333, so
+///    N_IVC = 3.41 N_APP
+///
+/// It is also useful to know that N / 2^15 = N_APP / (2^15 - CELL).
+/// In our case it means that N / 2^15 = N_APP / 7414, so if e.g.
+/// N_APP < 2170, N / 2^15 > 3 which means we can fit all 3N
+/// commitments in a single column!
+///
+/// I'm hopeful that this is the case, in which case the design will be as follows:
+///
+/// Layout under assumption that 3N < 2^15!
+///```text
+///           34      8    4        constϕ
+///                                   ϕ^i    same ϕs but
+///                                    r*ϕ^i  in 17 limbs  35*3 - BOT
+///         Input1    R75 R150  H1 H2          each         FEC ADDs
+///  1   |----------|----|----|-------|-|-|-|---|---|----------------|
+///      |   C_L1   |    |    |       |     |       |                |
+///      |   C_L2   |    |    |       |     |       |                |
+///      |   ...    |    |    |       |     |       |                |
+///   N  |----------|    |    |       |     |       |                |
+///      |   C_R1   |    |    |       |     |       |                |
+///      |   ...    |    |    |       |     |       |                |
+///  2N  |----------|    |    |       |     |       |                |
+///      |   C_O1   |    |    |       |     |       |                |
+///      |   ...    |    |    |       |     |       |                |
+///  3N  |----------|----|----|-------|-----|-------|----------------|
+///      |        default_instance?                                  |
+///      |                      ....... BOT FEC ADDs .......         |
+/// 2^15 |-----------------------------------------------------------|
+///```
+///
+///
+
+// NB: We can reuse hash constants.
+// TODO: Can we pass just one coordinate and sign (x, sign) instead of (x,y) for hashing?
+#[derive(Debug, Clone, PartialEq)]
+pub enum IVCColumn {
+    /// 2*17 15-bit limbs (two base field points)
+    Input(usize),
+    /// 2*4 75-bit limbs
+    InputRepacked75(usize),
+    /// 2*2 150-bit limbs
```

Comment:
do we need both 75 and 150?

Reply:
Yeah. 150 is as much as we can do for hashing, and 75bit is maximum that FEC ADD is working with... but it's cheap to have both, just a few extra rows. Not hundreds :)